Add support for generating and verifying digital signatures

paulw11 opened the pull request:
At present the signing function throws an exception if you try and sign too much data.  Ideally the function would generate an SHA1 digest of the data and sign that, but integration of CommonCrypto into Swift frameworks seems to be tricky. See - http://iosdeveloperzone.com/2014/10/03/using-commoncrypto-in-swift/

quentinlesceller commented:
Hi Paulw11,

Firs of all thank you for your contribution! 
Right now, I cannot merge your pull request as it is not secure to use without a hash:
- We can only sign messages up to a certain size.
- Possible forgery attacks.
- Reordering attacks.

Sources : [[1]](http://crypto.stackexchange.com/questions/12768/why-hash-the-message-before-signing-it-with-rsa), [[2]](http://crypto.stackexchange.com/questions/9896/how-does-rsa-signature-verification-work), [[3]](http://crypto.stackexchange.com/questions/2474/why-hash-or-salt-when-signing) and [[4]](http://crypto.stackexchange.com/questions/15295/why-the-need-to-hash-before-signing-small-data).

I think the best way for us to implement a digital signature in SwiftyRSA would be to implement a SHA1 function (like [this one in CryptoSwift ](https://github.com/krzyzanowskim/CryptoSwift/blob/master/Sources/CryptoSwift/SHA1.swift) which is based on this [Wiki pseudocode](https://en.wikipedia.org/wiki/SHA-1#SHA-1_pseudocode)).

Moreover the hash should be padded (see example [here](http://csrc.nist.gov/groups/ST/toolkit/documents/dss/RSAstatement_10-12-06.pdf)).

For now, implementing a digital signature with the hash (without padding) should be a good start.
Again thank you for your contribution. Adding digital signatures would be very useful as I didn't find any framework which integrates it.

If you make those changes I will be more than happy to merge them into SwiftyRSA.

paulw11 commented:
Would it be more sensible to include CryptoSwift as a dependency than to implement SHA1 in SwiftyRSA?

paulw11 commented:
I have added an SHA1 digest function using the CommonCrypto library and added PKCSSHA1 padding.  An integration test is needed to make sure it works once it is integrated with a project through the pod

quentinlesceller commented:
Ok this looks promising, just tried it (by building it myself) and it worked perfectly. However, Pod installation fails because of CommonCrypto.
I will review this and get back to you asap.

paulw11 commented:
This is the method I used for integrating CommonCrypto - http://stackoverflow.com/questions/25248598/importing-commoncrypto-in-a-swift-framework

ldiqual commented:
@paulw11 Great work, thanks a lot for bringing this feature to SwiftyRSA!

In addition to @quentinlesceller 's comments, I'd like to make sure we have the following before we merge:
 - Signing section the README that explains what are the available signing/verification methods in SwiftyRSA, and also documents the SHA1 part.
 - Add a line in CHANGELOG that describes your changes. Please link to the current pull request.
 - Swift + ObjC tests for class methods `signString, signData`, and both variants of `verifySignatureString, verifySignatureData`
 - Swift + ObjC tests of the instance methods counterpart.

I'm still kind of torn about the need of duplicating Swift / ObjC tests, but I've been bitten in the past by the lack of `@objc` that prevented the library from being used in a ObjC environment. Let me know if you guys know a better alternative to make sure everything works fine.

I also see a couple style issues that I'd like you to address, I'll comment inline for those. I think it should be enforced with Swiftlint, but for now let's do it manually.

Again, great work, this is an awesome feature to add and I particularly appreciate the time you took to include CommonCrypto since it seems like a pain to implement.

ldiqual commented on SwiftyRSA/NSData+SHA1.swift:
```diff
@@ -0,0 +1,22 @@
+    func SHA1() -> NSData {
+        var digest = [UInt8](count:Int(CC_SHA1_DIGEST_LENGTH), repeatedValue: 0)
+        
+        CC_SHA1(self.bytes, CC_LONG(self.length), &digest)
```
Do you really need `self` here?

paulw11 replied:
It is up to you.  My preference is to use `self` whenever referring to a property.  This is a habit from Objective-C where there is a difference between accessing the property and accessing the iVar directly.

ldiqual commented on SwiftyRSA/NSData+SHA1.swift:
```diff
@@ -0,0 +1,22 @@
+
+extension NSData {
+    func SHA1() -> NSData {
+        var digest = [UInt8](count:Int(CC_SHA1_DIGEST_LENGTH), repeatedValue: 0)
```
Space after `count:`

ldiqual commented on SwiftyRSA/SwiftyRSA.swift:
```diff
@@ -66,6 +67,43 @@ public class SwiftyRSA: NSObject {
+        return try rsa.signData(data, privateKey: key)
+    }
+    
+    public class func verifySignatureString(str: String, signature: String, publicKeyPEM: String) throws -> Void {
```
Shouldn't the verify methods return a boolean instead of void?
 - `true` => verification succeeds
 - `false` => verification fails
 - error => something unexpected happened

paulw11 replied:
That was the initial approach I took, however it seems that a function that returns Bool and also `throws` is invisible to Objective-C. The only way I could get the `verify...` methods to be visible to Objective C was to make them Void

ldiqual commented on SwiftyRSA/SwiftyRSA.swift:
```diff
@@ -170,6 +208,92 @@ public class SwiftyRSA: NSObject {
+    // Sign data with an RSA key
+    
+    public func signString(str: String, privateKey: SecKeyRef) throws -> String {
+        guard let data=str.dataUsingEncoding(NSUTF8StringEncoding) else {
```
Spaces around  `=`

ldiqual commented on SwiftyRSATests/SwiftyRSAObjcTests.m:
```diff
@@ -104,4 +104,27 @@ - (void)testDataEncryptDecrypt {
+    NSString* pubString = [TestUtils pemKeyStringWithName:@"swiftyrsa-public"];
+    NSString* privString = [TestUtils pemKeyStringWithName:@"swiftyrsa-private"];
+    
+    NSError *error;
```
Let's keep a consistent pointer style: `Type* variable`

ldiqual commented on SwiftyRSATests/SwiftyRSAObjcTests.m:
```diff
@@ -104,4 +104,27 @@ - (void)testDataEncryptDecrypt {
+    
+    NSString *signature = [SwiftyRSA signString:[data description] privateKeyPEM:privString error:&error];
+    
+    XCTAssert(error == nil);
```
I believe you can use XCTAssertNotNil

paulw11 commented:
`signString` and `signData` are both tested in the current Swift test suite. Since the class methods create an instance of `SwiftyRSA` and then invoke the corresponding instance method on that instance method, testing the class method tests the instance method.  I don't believe that specific instance method tests are required.

In fact, as `signString` and `verifySignatureString` convert the string to data and then call `signData` and `verifySignatureData` explicit tests of the `...Data` methods probably aren't needed either, but I have included them

paulw11 commented:
I have updated the readme and changelog files.  I also added documentation for the sign/verify methods in the docs directory - These were generated using Jazzy but are obviously incomplete as the encrypt/decrypt functions aren't documented

paulw11 commented:
I was able to have the `verify...` methods return a Boolean with the minimum amount of disruption to the Swift usage while still being compatible with Objective C

mwaterfall commented:
This is great guys, and perfect timing too as I'm just looking to work with digital signatures! Excellent stuff. Looking forward to seeing it merged in.

quentinlesceller commented:
Okay, I have looked at your pull request. First, naive question, why do you return a VerificationResult instead of a Bool ?

Also,  can you update your pull request for the latest version (0.2.1) ?

There is still an error with Cocoapod because CommonCrypto is missing. There should be a way by using `s.module_map = 'CommonCrypto/iphoneos.modulemap'` in SwiftyRSA.podspec but I did not find it yet (see [here](http://stackoverflow.com/questions/30608575/include-pods-bridging-header-to-build-settings-of-projects-target)) or maybe Cocoapod can automatically generate a bridging header for CommonCrypto ?
Also, you should modify this line in SwiftyRSA.podspec : `  s.source_files = "SwiftyRSA/*.swift"`.

I've tried some integration tests (carthage and manual build) and everything works great.
After you make those changes, we should be ready to merge.

ldiqual commented:
@quentinlesceller Please hold on this, I'd like to make sure we can't find a cleaner solution with booleans rather than having another object to document.

ldiqual commented:
@paulw11 @quentinlesceller I looked into this issue a bit more, and as Paul said there's no way to return a Swift primitive from a throwing function if you want to bridge it to ObjC. The `BooleanType` trick is pretty neat as there's an implicit conversion to Bool in swift, so I'm good with the `VerificationResult` addition.

@paulw11 Thank you for documenting your methods, it'll be useful if/when we want auto-generation documentation. However I don't think it's quite the time yet to add a `docs` section, especially since html is not very useful without a website. Let's try to reduce the scope of the PR. Would you mind removing the generated doc? As long as it's documented in the README I think we're fine.

ldiqual commented on CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 
+# https://github.com/TakeScoop/SwiftyRSA/pull/7
+
+ - Added digital signature creation & verification support
```
This should be in the master section, please copy the "style" of the other bullet points in there.

ldiqual commented on README.md:
````diff
@@ -53,6 +53,33 @@ let decryptedString = try! SwiftyRSA.decryptString(str, privateKeyPEM: pemString
+
+```
+// String
+try! SwitfyRSA.verifySignatureString(str, signature: sigString, publicKeyPEM: pemString)
````
Shouldn't this be `let success = try! SwitfyRSA.verifySignatureString(str, signature: sigString, publicKeyPEM: pemString)`? Same for the next lines.

paulw11 commented:
As I was afraid, integrating Common Crypto into the pod process is quite tricky.  There are a few options I can see.  One is to use a different implementation of SHA1 by declaring CryptoSwift as a dependency, but I did like the idea of using Apple's crypto library as it is more likely to be patched.  The second is to include the Apple headers into the SwiftyRSA bridging header like RNCryptor does - https://github.com/RNCryptor/RNCryptor but this requires the user to mess around with their bridging header when the integrate the Pod.  The third approach that I am still experimenting with is creating a separate podspec within SwiftyRSA that creates the CommonCrypto target

quentinlesceller commented:
Yeah that was I expected too. 
I'm writing a SHA1 function in Swift for SwiftyRSA, if it works we would be able to use your code without any modifications. 
But I also like the idea to uses Apple's library especially for things like crypto.
Don't know if it helps but there is a pod [CommonCrypto](https://cocoapods.org/?q=commoncrypto) and one especially made to use CommonCrypto's hash function: [SwiftDataHash](https://cocoapods.org/pods/SwiftDataHash).

paulw11 commented:
I took an alternative approach; I moved the SHA1 function into an Objective-C category.  This seems to work when I include the pod in a test project

quentinlesceller commented:
Okay, tried both integrating with Carthage and CocoaPod and everything seems to work.

However, it seems that your forked does not have the 0.2.1 tag ( not really a problem since we will probably update the version once your work is merged).
@ldiqual

paulw11 commented:
I just fetched the upstream again and got the new tag.  Not sure why it didn't come across when I did that the other day, but it doesn't want to push that tag to my repo, perhaps I need to make some other change in order to give it something to push.  It wont be a problem anyway, because none of my changes should be tagged 0.2.1 anyway

ldiqual commented:
Looking great! Thanks a lot @paulw11 for adding this feature! This is a great improvement. I'm gonna merge your changes now and tag this version as 0.3.0 after adding some linter rules. Also, @quentinlesceller thanks for taking time to review and test the changes. I'll go ahead and publish the missing tags on cocoapods.

@paulw11 Given the great work in this PR, I've sent you an invite to join the SwiftyRSA maintainers group -- no pressure to accept! Feel free to get back to me if you have any question.
